src/SegmentedArray.chpl

```diff
@@ -176,39 +176,44 @@ module SegmentedArray {
-         it is the difference between the src offset of the current segment ("left")
-         and the src index of the last byte in the previous segment (right - 1).
-       */
-      var srcIdx = makeDistArray(retBytes, int);
-      srcIdx = 1;
-      var diffs: [D] int;
-      diffs[D.low] = left[D.low]; // first offset is not affected by scan
-      diffs[{D.low+1..D.high}] = left[{D.low+1..D.high}] - (right[{D.low..D.high-1}] - 1);
-      // Set srcIdx to diffs at segment boundaries
-      forall (go, d) in zip(gatheredOffsets, diffs) with (var agg = newDstAggregator(int)) {
-        agg.copy(srcIdx[go], d);
-      }
-      srcIdx = + scan srcIdx;
-      // Now srcIdx has a dst-local copy of the source index and vals can be efficiently gathered
-      ref va = values.a;
-      forall (v, si) in zip(gatheredVals, srcIdx) with (var agg = newSrcAggregator(uint(8))) {
-        agg.copy(v, va[si]);
+      // For comm layer with poor small-message performance, use aggregation
+      // at the expense of memory. Otherwise, unorderedCopy is faster and smaller.
+      if !(CHPL_COMM == 'none' || CHPL_COMM == 'ugni') {
+        // Compute the src index for each byte in gatheredVals
+        /* For performance, we will do this with a scan, so first we need an array
+           with the difference in index between the current and previous byte. For
+           the interior of a segment, this is just one, but at the segment boundary,
+           it is the difference between the src offset of the current segment ("left")
+           and the src index of the last byte in the previous segment (right - 1).
+        */
+        if DEBUG { writeln("Using aggregation at the expense of memory"); stdout.flush(); }
+        var srcIdx = makeDistArray(retBytes, int);
+        srcIdx = 1;
+        var diffs: [D] int;
+        diffs[D.low] = left[D.low]; // first offset is not affected by scan
+        diffs[{D.low+1..D.high}] = left[{D.low+1..D.high}] - (right[{D.low..D.high-1}] - 1);
+        // Set srcIdx to diffs at segment boundaries
+        forall (go, d) in zip(gatheredOffsets, diffs) with (var agg = newDstAggregator(int)) {
+          agg.copy(srcIdx[go], d);
+        }
+        srcIdx = + scan srcIdx;
+        // Now srcIdx has a dst-local copy of the source index and vals can be efficiently gathered
+        ref va = values.a;
+        forall (v, si) in zip(gatheredVals, srcIdx) with (var agg = newSrcAggregator(uint(8))) {
+          agg.copy(v, va[si]);
+        }
+      } else {
+        if DEBUG { writeln("Using unorderedCopy"); stdout.flush(); }
+        ref va = values.a;
+        // Copy string data to gathered result
+        forall (go, gl, idx) in zip(gatheredOffsets, gatheredLengths, iv) {
+          for pos in 0..#gl {
+            // Note: do not replace this unorderedCopy with aggregation
+            use UnorderedCopy;
+            unorderedCopy(gatheredVals[go+pos], va[oa[idx]+pos]);
```

Comment:
Is the reason for the more complicated aggregation above instead just using this code with aggregation because the remoteness of Src/Dst isn't consistent? i.e. if there was a general purpose CopyAggregator, would that allow you to use this simpler loop?

Reply:
Yes, this is a situation where either, both, or neither Src/Dst could be remote. If there were a general purpose `CopyAggregator` that figured out what to do based on the remoteness of Src and Dst, I would definitely use it here for a much simpler loop. The performance would probably improve, and the memory usage might even go down (the existing code can use up to 8x the input data in scratch memory).